README.md
# toshl-python ![License MIT](https://go-shields.herokuapp.com/license-MIT-blue.png)

[![Travis-CI Status](https://secure.travis-ci.org/andreagrandi/toshl-python.png?branch=master)](http://travis-ci.org/#!/andreagrandi/toshl-python)

Python client library for Toshl API

## Getting started with the library

To use the library, you first need to obtain a **token** from Toshl API.
Once you have a token you need to import the library and create an instance of the client:

```
from toshl.client import ToshlClient
client = ToshlClient('xxx-xxxxx-xxx-xxxxxx-xxxxxx-xxx-xxxxxx')
```

Every resource exposes the following **methods**:

* list()
* search()
* get()
* create()
* update()
* delete()

To use each resource you will need to create an instance of them, passing the client as parameter in the constructor.

**Note:** at the moment not all the methods and not all the resources have been implemented.

Here you can find a couple of examples, but for the complete documentation you can have a look at the official website https://developer.toshl.com/docs/

### Accounts

```
from toshl.client import ToshlClient, Account

client = ToshlClient('xxx-xxxxx-xxx-xxxxxx-xxxxxx-xxx-xxxxxx')
account = Account(client)

# list all accounts
account.list()

# search for a specific account
account.search('Test Account')
```

### Categories

```
from toshl.client import ToshlClient, Category

client = ToshlClient('xxx-xxxxx-xxx-xxxxxx-xxxxxx-xxx-xxxxxx')
category = Category(client)

# list all categories
category.list()

# search for a specific category
category.search('Test Category')
```

### Entries

```
from toshl.client import ToshlClient, Entry

client = ToshlClient('xxx-xxxxx-xxx-xxxxxx-xxxxxx-xxx-xxxxxx')
entry = Entry(client)

# create an Entry
json_payload = {
    'amount': -123.68,
    'currency': {
        'code': 'GBP'
    },
    'date': '2016-04-07',
    'account': 'abcd1234',
    'category': 'category-001'
}

response = entry.create(json_payload)
```

## Copyright Note
**Toshl** and its logos, design, text, graphics, and other files, and the selection arrangement and organization thereof, are owned by Toshl.
This is a 3rd party code and I'm not affiliated nor I work for Toshl.


CHANGELOG.md
# toshl-python - Python client library for Toshl API

## CHANGELOG

### v0.0.4

- Add get() method to Account
- Add create() method to Account
- Add update() method to Account
- Add delete() method to Account
- Add move() method to Account
- Add reorder() method to Account
- Add merge() method to Account
- Add proper testing for ToshlException
- Improve string representation of ToshlException
- refactoring: each entity is in a separate module


tests/test_entry.py
import mock
from unittest import TestCase
from toshl.client import ToshlClient
from toshl.entry import Entry


class TestEntry(TestCase):
    def test_entry_init(self):
        client = ToshlClient('abcd1234')
        entry = Entry(client)
        assert entry.client == client

    @mock.patch('toshl.client.requests.request')
    def test_create_entry_successful(self, mock_request):
        mock_response = mock.Mock()
        mock_response.status_code = 201
        mock_response.headers = {'Location': '/entries/1'}
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        entry = Entry(client)

        json_payload = {
            'amount': -123.68,
            'currency': {
                'code': 'GBP'
            },
            'date': '2016-04-07',
            'account': 'abcd1234',
            'category': 'category-001'
        }

        response = entry.create(json_payload)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='POST', params=None, url='https://api.toshl.com/entries',
            json=json_payload)
        assert response == '1'


tests/test_client.py
# -*- coding: utf-8 -*-
import mock
from unittest import TestCase
from toshl.client import ToshlClient
from toshl.exceptions import ToshlException


class TestClient(TestCase):
    def setUp(self):
        self.success_mocked_response = mock.Mock()
        self.success_mocked_response.status_code = 200

    def test_client_init(self):
        client = ToshlClient('abcd1234')
        assert client._token == 'abcd1234'

    def test_client_has_base_url(self):
        client = ToshlClient('abcd1234')
        base_api_url = getattr(client, 'BASE_API_URL', None)
        assert base_api_url is not None

    @mock.patch('toshl.client.requests.request')
    def test_client_can_make_request(self, mock_request):
        mock_request.return_value = self.success_mocked_response
        client = ToshlClient('abcd1234')
        client._make_request('/me', 'GET')
        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/me')

    @mock.patch('toshl.client.requests.request')
    def test_client_can_make_request_with_params(self, mock_request):
        mock_request.return_value = self.success_mocked_response
        client = ToshlClient('abcd1234')
        client._make_request('/me', 'GET', params={'a': 'foo1', 'b': 'foo2'})
        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params={'a': 'foo1', 'b': 'foo2'},
            url='https://api.toshl.com/me')

    @mock.patch('toshl.client.requests.request')
    def test_client_can_make_request_with_utf8_params(self, mock_request):
        mock_request.return_value = self.success_mocked_response
        client = ToshlClient('abcd1234')
        client._make_request(
            '/me', 'GET', params={'a': u'foo1', 'b': u'£123'})
        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params={'a': u'foo1', 'b': u'\xa3123'},
            url='https://api.toshl.com/me')

    @mock.patch('toshl.client.requests.request')
    def test_client_set_auth_token_correctly(self, mock_request):
        mock_request.return_value = self.success_mocked_response
        client = ToshlClient('AAABBBCCCDDD')
        client._make_request('/me', 'GET', params={'a': 'foo1', 'b': 'foo2'})
        assert mock_request.call_args[1]['headers']['Authorization'] == \
            'Bearer AAABBBCCCDDD'

    @mock.patch('toshl.client.requests.request')
    def test_client_request_raises_exception_on_400(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            'error_id': 'exception_id',
            'description': 'Exception description'
        }
        mock_response.json.return_value = expected_dict
        mock_response.status_code = 400
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')

        with self.assertRaises(ToshlException) as ex:
            client._make_request('/me', 'GET')

        assert ex.exception.status_code == 400
        assert ex.exception.error_id == 'exception_id'
        assert ex.exception.error_description == 'Exception description'
        assert ex.exception.extra_info is None

    @mock.patch('toshl.client.requests.request')
    def test_client_request_raises_exception_with_extra_info(
            self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            'error_id': 'exception_id',
            'description': 'Exception description',
            "fields": [
                {
                    "field": "amount",
                    "error": "Amount cannot be zero."
                },
                {
                    "field": "category",
                    "error": "Please select at least one category."
                }
            ]
        }
        mock_response.json.return_value = expected_dict
        mock_response.status_code = 400
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')

        with self.assertRaises(ToshlException) as ex:
            client._make_request('/me', 'GET')

        assert ex.exception.status_code == 400
        assert ex.exception.error_id == 'exception_id'
        assert ex.exception.error_description == 'Exception description'
        assert ex.exception.extra_info == [
            {
                "field": "amount",
                "error": "Amount cannot be zero."
            },
            {
                "field": "category",
                "error": "Please select at least one category."
            }
        ]


tests/test_account.py
import mock
from unittest import TestCase
from toshl.client import ToshlClient
from toshl.account import Account
from toshl.exceptions import ToshlException


class TestAccount(TestCase):
    def test_account_init(self):
        client = ToshlClient('abcd1234')
        account = Account(client)
        assert account.client == client

    @mock.patch('toshl.client.requests.request')
    def test_list_accounts_successful(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            "id": "42",
            "name": "Account Test",
            "balance": 3000,
            "initial_balance": 3000,
            "currency": {
                "code": "USD",
                "rate": 1,
                "fixed": False
            },
            "median": {
                "expenses": 55,
                "incomes": 1300
            },
            "status": "active",
            "order": 0,
            "modified": "2012-09-04T13:55:15Z",
            "goal": {
                "amount": 63570,
                "start": "2013-07-01",
                "end": "2015-07-01"
            }
        }
        mock_response.json.return_value = expected_dict
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.list()

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/accounts')
        assert response == [expected_dict]

    @mock.patch('toshl.client.requests.request')
    def test_search_accounts_successful_multiple_accounts(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = [
            {
                "id": "42",
                "name": "Account Test",
                "balance": 3000
            },
            {
                "id": "123",
                "name": "Test Found",
                "balance": 22000
            }
        ]

        mock_response.json.return_value = expected_dict
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        account_found = account.search('Test Found')

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/accounts')
        assert account_found is not None
        assert account_found == '123'

    @mock.patch('toshl.client.requests.request')
    def test_search_accounts_successful_single_account(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            "id": "123",
            "name": "Test Found",
            "balance": 22000
        }

        mock_response.json.return_value = expected_dict
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        account_found = account.search('Test Found')

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/accounts')
        assert account_found is not None
        assert account_found == '123'

    @mock.patch('toshl.client.requests.request')
    def test_search_accounts_not_found(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            "id": "123",
            "name": "Test Found",
            "balance": 22000
        }

        mock_response.json.return_value = expected_dict
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        account_found = account.search('Not Found')

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/accounts')
        assert account_found is None

    @mock.patch('toshl.client.requests.request')
    def test_get_account_successful(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            "id": "42",
            "name": "Account Test",
            "balance": 3000,
            "initial_balance": 3000,
            "currency": {
                "code": "USD",
                "rate": 1,
                "fixed": False
            },
            "median": {
                "expenses": 55,
                "incomes": 1300
            },
            "status": "active",
            "order": 0,
            "modified": "2012-09-04T13:55:15Z",
            "goal": {
                "amount": 63570,
                "start": "2013-07-01",
                "end": "2015-07-01"
            }
        }
        mock_response.json.return_value = expected_dict
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.get('42')

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='GET', params=None, url='https://api.toshl.com/accounts/42')
        assert response == expected_dict

    @mock.patch('toshl.client.requests.request')
    def test_account_not_found_raises_exception(self, mock_request):
        mock_response = mock.Mock()
        expected_dict = {
            'error_id': 'error.object.not_found',
            'description': 'Object with id 111 not found.'
        }
        mock_response.json.return_value = expected_dict
        mock_response.status_code = 404
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')

        with self.assertRaises(ToshlException) as ex:
            client._make_request('/accounts/111', 'GET')

        assert ex.exception.status_code == 404
        assert ex.exception.error_id == 'error.object.not_found'
        assert ex.exception.error_description == 'Object with id 111 not found.'
        assert ex.exception.extra_info is None

    @mock.patch('toshl.client.requests.request')
    def test_create_account_successful(self, mock_request):
        mock_response = mock.Mock()
        mock_response.status_code = 201
        mock_response.headers = {'Location': '/accounts/1'}
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account_client = Account(client)

        json_payload = {
            'name': 'Test Account',
            'currency': {
                'code': 'GBP'
            }
        }

        response = account_client.create(json_payload)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='POST', params=None, url='https://api.toshl.com/accounts',
            json=json_payload)
        assert response == '1'

    @mock.patch('toshl.client.requests.request')
    def test_update_account_successful(self, mock_request):
        json_payload = {
            'name': 'Test Account',
            'currency': {
                'code': 'GBP'
            },
            'extra': {
                'test': 'foo'
            }
        }

        mock_response = mock.Mock()
        mock_response.status_code = 200
        mock_response.json.return_value = json_payload
        mock_response.headers = {'Location': '/accounts/1'}
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account_client = Account(client)

        response = account_client.update('1', json_payload)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='PUT', params=None, url='https://api.toshl.com/accounts/1',
            json=json_payload)
        assert response == json_payload

    @mock.patch('toshl.client.requests.request')
    def test_delete_account_successful(self, mock_request):
        mock_response = mock.Mock()
        mock_response.json.return_value = ''
        mock_response.status_code = 204
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.delete('1')

        mock_request.assert_called_once_with(
            headers={'Authorization': 'Bearer abcd1234'},
            method='DELETE',
            params=None, url='https://api.toshl.com/accounts/1')
        assert response.json() == ''

    @mock.patch('toshl.client.requests.request')
    def test_move_account_successful(self, mock_request):
        json_payload = {
            'position': 2
        }

        mock_response = mock.Mock()
        mock_response.json.return_value = ''
        mock_response.status_code = 204
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.move('1', 2)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='POST',
            params=None, url='https://api.toshl.com/accounts/1/move',
            json=json_payload)
        assert response.json() == ''

    @mock.patch('toshl.client.requests.request')
    def test_reorder_account_successful(self, mock_request):
        accounts_list = ['3', '1', '2']

        json_payload = {
            'order': accounts_list
        }

        mock_response = mock.Mock()
        mock_response.json.return_value = ''
        mock_response.status_code = 204
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.reorder(accounts_list)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='POST',
            params=None, url='https://api.toshl.com/accounts/reorder',
            json=json_payload)
        assert response.json() == ''

    @mock.patch('toshl.client.requests.request')
    def test_merge_account_successful(self, mock_request):
        accounts_list = ['1', '2']
        dest_account = '2'

        json_payload = {
            'accounts': accounts_list,
            'account': dest_account
        }

        mock_response = mock.Mock()
        mock_response.json.return_value = ''
        mock_response.status_code = 204
        mock_request.return_value = mock_response

        client = ToshlClient('abcd1234')
        account = Account(client)
        response = account.merge(accounts_list, dest_account)

        mock_request.assert_called_once_with(
            headers={
                'Authorization': 'Bearer abcd1234',
                'Content-Type': 'application/json'
            },
            method='POST',
            params=None, url='https://api.toshl.com/accounts/merge',
            json=json_payload)
        assert response.json() == ''


toshl/client.py
import requests
from .exceptions import ToshlException


class ToshlClient(object):
    BASE_API_URL = 'https://api.toshl.com'

    def __init__(self, token):
        self._token = token

    def _make_request(
            self, api_resource, method='GET', params=None, **kwargs):
        """
        Shortcut for a generic request to the Toshl API
        :param url: The URL resource part
        :param method: REST method
        :param parameters: Querystring parameters
        :return: requests.Response
        """
        if kwargs.get('json'):
            headers = {
                'Authorization': 'Bearer {}'.format(self._token),
                'Content-Type': 'application/json'
            }
        else:
            headers = {
                'Authorization': 'Bearer {}'.format(self._token)
            }

        response = requests.request(
            method=method,
            url='{0}{1}'.format(self.BASE_API_URL, api_resource),
            headers=headers,
            params=params,
            **kwargs
        )

        if response.status_code >= 400:
            error_response = response.json()

            raise(ToshlException(
                status_code=response.status_code,
                error_id=error_response['error_id'],
                error_description=error_response['description'],
                extra_info=error_response.get('fields')))

        return response

    def _list_response(self, response):
        """
        This method check if the response is a dict and wrap it into a list.
        If the response is already a list, it returns the response directly.
        This workaround is necessary because the API doesn't return a list
        if only one item is found.
        """
        if type(response) is list:
            return response
        if type(response) is dict:
            return [response]

    def _parse_location_header(self, response):
        return response.headers['Location'].split('/')[-1:][0]


toshl/account.py
class Account(object):
    def __init__(self, client):
        self.client = client

    def list(self):
        """
        Return a list of Accounts from Toshl for the current user
        """
        response = self.client._make_request('/accounts')
        response = response.json()
        return self.client._list_response(response)

    def search(self, account_name):
        """
        Get a list of all the Accounts for the current user and return the ID
        of the one with the specified name.
        """
        accounts = self.list()
        for a in accounts:
            if a['name'] == account_name:
                return a['id']

    def get(self, account_id):
        """
        Return a specific account given its ID
        """
        response = self.client._make_request('/accounts/{0}'.format(account_id))
        return response.json()

    def create(self, json_payload):
        response = self.client._make_request(
            '/accounts', 'POST', json=json_payload)
        if response.status_code == 201:
            return self.client._parse_location_header(response)

    def update(self, account_id, json_payload):
        response = self.client._make_request(
            '/accounts/{0}'.format(account_id), 'PUT', json=json_payload)
        if response.status_code == 200:
            return response.json()

    def delete(self, account_id):
        return self.client._make_request(
            '/accounts/{0}'.format(account_id), 'DELETE')

    def move(self, account_id, position):
        json_payload = {
            'position': position
        }

        return self.client._make_request(
            '/accounts/{0}/move'.format(account_id), 'POST', json=json_payload)

    def reorder(self, accounts_list):
        json_payload = {
            'order': accounts_list
        }

        return self.client._make_request(
            '/accounts/reorder', 'POST', json=json_payload)

    def merge(self, accounts_list, dest_account):
        json_payload = {
            'accounts': accounts_list,
            'account': dest_account
        }

        return self.client._make_request(
            '/accounts/merge', 'POST', json=json_payload)
